core/unix/rseq_linux.c

```diff
@@ -0,0 +1,377 @@
+                ++ptrs;
+            }
+            break;
+        }
+        ++sec_hdr;
+    }
+    if (!found_array) {
+        sec_hdr = sec_hdr_start;
+        for (i = 0; i < elf_hdr->e_shnum; i++) {
+#define RSEQ_SEC_NAME "__rseq_cs"
+#define RSEQ_OLD_SEC_NAME "__rseq_table"
+            if (strcmp(strtab + sec_hdr->sh_name, RSEQ_SEC_NAME) == 0 ||
+                strcmp(strtab + sec_hdr->sh_name, RSEQ_OLD_SEC_NAME) == 0) {
+                /* There may be padding at the start of the section, so ensure we skip
+                 * over it.  We're reading the loaded data, not the file, so it will
+                 * always be aligned.
+                 */
+#define RSEQ_CS_ALIGNMENT (4 * sizeof(__u64))
+                struct rseq_cs *array = (struct rseq_cs *)ALIGN_FORWARD(
+                    sec_hdr->sh_addr + load_offs, RSEQ_CS_ALIGNMENT);
+                int j;
+                for (j = 0; j < sec_hdr->sh_size / sizeof(*array); ++j) {
+                    /* We require that the table is loaded.  If not, bail. */
+                    if (array < (struct rseq_cs *)ma->start ||
+                        array > (struct rseq_cs *)ma->end) {
+                        REPORT_FATAL_ERROR_AND_EXIT(
+                            RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                            get_application_pid(),
+                            RSEQ_SEC_NAME " is not in a loaded segment");
+                        ASSERT_NOT_REACHED();
+                    }
+                    rseq_process_entry(array, entry_offs);
+                    ++array;
+                }
+                break;
+            }
+            ++sec_hdr;
+        }
+    }
+    res = true;
+rseq_process_module_cleanup:
+    if (str_size != 0)
+        os_unmap_file(str_map, str_size);
+    if (sec_size != 0)
+        os_unmap_file(sec_map, sec_size);
+    if (fd != INVALID_FILE)
+        os_close(fd);
+    DODEBUG({
+        if (!res) {
+            const char *name = GET_MODULE_NAME(&ma->names);
+            if (name == NULL)
+                name = "(null)";
+            LOG(GLOBAL, LOG_INTERP | LOG_VMAREAS, 2,
+                "%s: error looking for rseq table in %s\n", __FUNCTION__, name);
+            if (strstr(name, "linux-vdso.so") == NULL) {
+                SYSLOG_INTERNAL_WARNING_ONCE(
+                    "Failed to identify whether a module has an rseq table");
+            }
+        }
+    });
+    return res;
+}
+
+/* Restartable sequence region identification.
+ *
+ * To avoid extra overhead going to disk to read section headers, we delay looking
+ * for rseq data until the app invokes an rseq syscall (or on attach we see a thread
+ * that has rseq set up).  We document that we do not handle the app using rseq
+ * regions for non-rseq purposes, so we do not need to flush the cache here.
+ */
+void
+rseq_locate_rseq_regions(void)
+{
+    if (rseq_enabled)
+        return;
+    d_r_mutex_lock(&rseq_trigger_lock);
```

Comment:
Is this lock functionally needed, can't two threads search, is there any resource conflict? Probably better to have it to be safe? Maybe short comment somewhere?

Reply:
No, two threads can't search all of the modules simultaneously: this is a global operation, delayed until the app invokes rseq.  It requires synchronization in case two threads hit the lazy trigger at once.  I thought it was relatively clear from the existing comments and code (it walks the global module list) but I can add another comment.